File: go/tree.go
package brackettree

import (
	"fmt"
	"strings"
)

// NewBracketTree parses bracket tree string and return the root node object of the tree
func NewBracketTree(treeString string) (*Node, error) {
	return parseTreeNodesFromRuneArray([]rune(treeString), 0, len(treeString))
}

type Node struct {
	Value    string
	Children []*Node
}

func (n Node) BracketRepresentation() string {
	str := n.Value
	for _, c := range n.Children {
		str += fmt.Sprintf("(%s)", c.BracketRepresentation())
	}
	return str
}

func (n *Node) Find(name string) *Node {
	if n.Value == name {
		return n
	}
	for _, c := range n.Children {
		if n := c.Find(name); n != nil {
			return n
		}
	}
	return nil
}

func (n Node) CountLeaves() int {
	// leaf nodes have 0 children
	if len(n.Children) == 0 {
		return 1
	}
	leaves := 0
	for _, c := range n.Children {
		leaves += c.CountLeaves()
	}
	return leaves
}

func (n *Node) AddChild(childTreeString string) error {
	node, err := parseTreeNodesFromRuneArray([]rune(childTreeString), 0, len(childTreeString))
	if err != nil {
		return err
	}
	n.Children = append(n.Children, node)
	return nil
}

// indexOfClosingBracket is a helper function returning an index of the closing bracket
func indexOfClosingBracket(treeRune []rune, startParenthesisIndex int) (int, error) {
	if treeRune[startParenthesisIndex] != '(' {
		return -1, fmt.Errorf(fmt.Sprintf("expected %q at index %v, but got %q", '(', startParenthesisIndex, treeRune[startParenthesisIndex]))
	}
	bracketCounter := 1
	for i := startParenthesisIndex + 1; i < len(treeRune); i++ {
		if treeRune[i] == '(' {
			bracketCounter++
		}
		if treeRune[i] == ')' {
			bracketCounter--
		}
		if bracketCounter == 0 {
			return i, nil
		}
	}
	return -1, fmt.Errorf("ran till the end, but could not find the ending bracket")
}

// parseTreeNodesFromRuneArray is a helper recursive function that takes an array of runes as input and returns the tree structure - object Node
func parseTreeNodesFromRuneArray(treeRune []rune, treeStartIndex int, treeEndIndex int) (*Node, error) {
	i := treeStartIndex
	// iterate through the rune array to find the node's value
	var sb strings.Builder
	for i < treeEndIndex && treeRune[i] != '(' && treeRune[i] != ')' {
		sb.WriteRune(treeRune[i])
		i++
	}
	node := &Node{Value: sb.String()}
	if node.Value == "" {
		return nil, fmt.Errorf(fmt.Sprintf("node value at index %v should not be empty", treeStartIndex))
	}

	// iterate through children and recursively call the function to parse children trees
	childTreeStartIndex := i
	for childTreeStartIndex < treeEndIndex {
		if treeRune[childTreeStartIndex] != '(' {
			return nil, fmt.Errorf("expected %q but got %q, failed at index %v", '(', treeRune[childTreeStartIndex], childTreeStartIndex)
		}
		closingBracketIndex, err := indexOfClosingBracket(treeRune, childTreeStartIndex)
		if err != nil {
			return nil, err
		}
		childTree, err := parseTreeNodesFromRuneArray(treeRune, childTreeStartIndex+1, closingBracketIndex)
		if err != nil {
			return nil, err
		}
		node.Children = append(node.Children, childTree)
		childTreeStartIndex = closingBracketIndex + 1
	}
	return node, nil
}

File: rust/src/tree.rs
use std::fmt;

#[derive(Debug)]
pub struct Node {
    pub name: String,
    pub children: Vec<Node>,
}

impl Node {
    pub(crate) fn add_child_node(&mut self, child: Node) {
        self.children.push(child);
    }

    pub fn add_child(&mut self, child_string: String) -> Result<Node, ParseError> {
        Err(ParseError { message: String::from("not implemented") })
    }

    pub fn count_leaves(&self) -> i32 {
        if self.children.is_empty() {
            return 1;
        }
        let mut sum: i32 = 0;
        for c in &self.children {
            sum += c.count_leaves();
        }
        sum
    }

    pub fn find(&self, name: &str) -> Result<Option<&Node>, ParseError> {
        if self.name == name {
            return Ok(Some(self));
        }
        for c in &self.children {
            let k = c.find(name)?;
            if k.is_some() {
                return Ok(k);
            }
        }
        Ok(None)
    }

    pub fn to_bracket_string(&self) -> String {
        let mut bracket_string = self.name.to_string();
        for c in &self.children {
            bracket_string.push_str("(");
            bracket_string.push_str(&c.to_bracket_string());
            bracket_string.push_str(")");
        }
        bracket_string
    }
}


#[derive(Debug, Clone)]
pub struct ParseError {
    pub message: String
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Parse Error: {}", self.message)
    }
}

pub fn parse(bracket_tree: &str) -> Result<Node, ParseError> {
    let char_vec: Vec<char> = bracket_tree.chars().collect();
    parse_bracket_string_to_node(&char_vec, 0, bracket_tree.len())
}

fn get_closing_bracket_index(bracket_tree: &[char], start_index: usize) -> Result<usize, ParseError> {
    if bracket_tree[start_index] != '(' {
        return Err(ParseError { message: format!("expected '{}' but found '{}' at index {}", "(", bracket_tree[start_index], start_index) });
    }

    let mut bracket_counter = 1;
    for (i, ch) in bracket_tree.iter().enumerate().skip(start_index + 1) {
        match ch {
            '(' => bracket_counter += 1,
            ')' => bracket_counter -= 1,
            _ => {}
        }
        if bracket_counter == 0 {
            return Ok(i);
        }
    }

    Err(ParseError { message: String::from("reached the end of string and could not find the matching closing bracket") })
}

fn parse_bracket_string_to_node(bracket_tree: &[char], start_index: usize, end_index: usize) -> Result<Node, ParseError> {
    let mut name = String::new();
    let mut child_start_index = start_index;
    while child_start_index < end_index && bracket_tree[child_start_index] != '(' {
        name.push(bracket_tree[child_start_index]);
        child_start_index += 1;
    }
    if name.is_empty() {
        return Err(ParseError { message: format!("node name of the tree starting at index {} is empty", start_index) });
    }

    let mut children = vec![];
    while child_start_index < end_index {
        let child_end_index = get_closing_bracket_index(bracket_tree, child_start_index)?;
        let child = parse_bracket_string_to_node(bracket_tree, child_start_index + 1, child_end_index)?;
        children.push(child);
        child_start_index = child_end_index + 1;
    }

    Ok(Node { name, children })
}